src/java.base/share/classes/jdk/internal/classfile/impl/DirectClassBuilder.java

```diff
@@ -197,26 +198,20 @@ else if ((flags & ClassFile.ACC_MODULE) == 0 && !"java/lang/Object".equals(thisC
         attributes.writeTo(tail);
 
         // Now we have to append the BSM, if there is one
-        boolean written = constantPool.writeBootstrapMethods(tail);
-        if (written) {
+        if (constantPool.writeBootstrapMethods(tail)) {
             // Update attributes count
             tail.patchU2(attributesOffset, attributes.size() + 1);
         }
 
         // Now we can make the head
-        head.writeInt(ClassFile.MAGIC_NUMBER);
-        head.writeU2(minorVersion);
-        head.writeU2(majorVersion);
+        head.writeLong(((long) ClassFile.MAGIC_NUMBER) << 32 | minorVersion << 16 | majorVersion);
```

Comment:
`minorVersion` needs to be cast to a `long` first, as otherwise when the MSB is set after the shift, then it’ll overwrite the magic number with all 1s.

```suggestion
        head.writeLong(((long) ClassFile.MAGIC_NUMBER) << 32 | ((long) minorVersion) << 16 | majorVersion);
```

Reply:
For clarity, I think using Integer.toUnsignedLong on the result is better

Reply:
If minorVersion > 0xFFFF, the result of using Integer.toUnsignedLong will be wrong. I guess this is the reason why the previous version build failed

Reply:
Should we add a `& 0xFFFF` on `majorVersion` to defend against `majorVersion > 0xFFFF` too?